README.md
## 一、简介

1. `Transmission`是一款在`Linux`系统上非常流行的`BT`客户端。
2. `Transmission`一键安装包使用`Linux Shell`语言编写，用于在`Linux`操作系统上一键安装`Transmission`。

## 二、下载地址

1. 方法一：通过Git下载（推荐）

    	git clone git://github.com/wangyan/trans.git
    	cd trans && ./install.sh

2. 方法二：直接下载已打包版本

    	wget http://wangyan.org/download/src/trans-latest.tar.gz
    	tar -zxf trans-*.tar.gz
    	cd trans && ./install.sh

## 三、安装步骤

1. 输入`Transmission`远程登录用户名。
2. 输入`Transmission`远程登录密码。
3. 输入`Transmission`下载目录，默认目录是`/root/Downloads`
4. 按任意键开始安装，可以按+c退出。

## 四、使用说明

1. 输入`/etc/init.d/transmission start`启动
2. 输入`/etc/init.d/transmission stop`停止
3. 输入`/etc/init.d/transmission restart`重启

## 五、扩展资源

> http://www.transmissionbt.com/resources/

例如：[transmisson-remote-gui 图形客户端](http://code.google.com/p/transmisson-remote-gui/)、  [Transdroid 手机android客户端](http://www.transdroid.org/about/)

## 六、注意事项

1. 修改配置文件需要重启
2. `Transmission`配置文件路径是`/usr/local/transmission/etc/settings.json.bak`    
    注意：请不要遗漏了`bak`，请不要直接修改`settings.json`文件。

## 七、联系方式

> 如果安装出错，请将安装目录下的 `log.txt` 文件提交给我处理。   
>   
> Email: [WangYan#188.com](WangYan#188.com) （推荐）  
> Gtalk: [myidwy#gmail.com](myidwy#gmail.com)  
> Q Q群：[138082163](http://qun.qq.com/#jointhegroup/gid/138082163)  
> Twitter：[@wang_yan](https://twitter.com/wang_yan)  
> Home Page: [WangYan Blog](http://wangyan.org/blog)  


Changelog
一、当前版本

version 0.1.10.28.03
- 修复64位系统transmission编译错误

二、更新历史

v0.1.10.28.02
- README新增使用说明

v0.1.10.05.01
- 首次提交


install.sh
#! /bin/bash
#====================================================================
# install.sh
#
# Transmission Auto Install Script
#
# Copyright (c) 2012, WangYan <WangYan@188.com>
# All rights reserved.
# Distributed under the GNU General Public License, version 3.0.
#
# Intro: http://wangyan.org/blog/transmission.html
#
#====================================================================

if [ $(id -u) != "0" ]; then
    clear && echo "Error: You must be root to run this script!"
    exit 1
fi

TRANS_PATH=`pwd`
if [ `echo $TRANS_PATH | awk -F/ '{print $NF}'` != "trans" ]; then
	clear && echo "Please enter trans script path:"
	read -p "(Default path: ${TRANS_PATH}/trans):" TRANS_PATH
	[ -z "$TRANS_PATH" ] && TRANS_PATH=$(pwd)/trans
	cd $TRANS_PATH/
fi

DISTRIBUTION=`awk 'NR==1{print $1}' /etc/issue`

if echo $DISTRIBUTION | grep -Eqi '(Red Hat|CentOS|Fedora|Amazon)';then
    PACKAGE="rpm"
elif echo $DISTRIBUTION | grep -Eqi '(Debian|Ubuntu)';then
    PACKAGE="deb"
else
    if cat /proc/version | grep -Eqi '(redhat|centos)';then
        PACKAGE="rpm"
    elif cat /proc/version | grep -Eqi '(debian|ubuntu)';then
        PACKAGE="deb"
    else
        echo "Please select the package management! (rpm/deb)"
        read -p "(Default: rpm):" PACKAGE
        if [ -z "$PACKAGE" ]; then
            PACKAGE="rpm"
        fi
        if [[ "$PACKAGE" != "rpm" && "$PACKAGE" != "deb" ]];then
            echo -e "\nNot supported linux distribution!"
            echo "Please contact me! WangYan <WangYan@188.com>"
            exit 0
        fi
    fi
fi

[ -r "$TRANS_PATH/fifo" ] && rm -rf $TRANS_PATH/fifo
mkfifo $TRANS_PATH/fifo
cat $TRANS_PATH/fifo | tee $TRANS_PATH/log.txt &
exec 1>$TRANS_PATH/fifo
exec 2>&1

/bin/bash ${TRANS_PATH}/${PACKAGE}.sh

sed -i '/password/d' $TRANS_PATH/log.txt
rm -rf $TRANS_PATH/fifo


rpm.sh
#!/bin/bash
PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin
export PATH

if [ $(id -u) != "0" ]; then
	printf "Error: You must be root to run this script!"
	exit 1
fi

TRANS_PATH=`pwd`
if [ `echo $TRANS_PATH | awk -F/ '{print $NF}'` != "trans" ]; then
	clear && echo "Please enter trans script path:"
	read -p "(Default path: ${TRANS_PATH}/trans):" TRANS_PATH
	[ -z "$TRANS_PATH" ] && TRANS_PATH=$(pwd)/trans
	cd $TRANS_PATH/
fi

clear
echo "#############################################################"
echo "# Transmission Auto Install Shell Script"
echo "# Env: Redhat/CentOS"
echo "# Intro: http://wangyan.org/blog/transmission.html"
echo "# Version: $(awk '/version/{print $2}' $TRANS_PATH/Changelog)"
echo "#"
echo "# Copyright (c) 2012, WangYan <WangYan@188.com>"
echo "# All rights reserved."
echo "# Distributed under the GNU General Public License, version 3.0."
echo "#"
echo "#############################################################"
echo ""

echo "Please enter remote login username:"
read -p "(Default username: demo):" USERNAME
if [ -z $USERNAME ]; then
	USERNAME="demo"
fi
echo "---------------------------"
echo "login username = $USERNAME"
echo "---------------------------"
echo ""

echo "Please enter remote login password:"
read -p "(Default password: demo):" PASSWORD
if [ -z $PASSWORD ]; then
	PASSWORD="demo"
fi
echo "---------------------------"
echo "login password = $PASSWORD"
echo "---------------------------"
echo ""

echo "Please enter the download dir:"
read -p "(Default download dir: /root/Downloads):" DLDIR
if [ -z $DLDIR ]; then
	DLDIR="/root/Downloads"
fi
echo "---------------------------"
echo "download dir=$DLDIR"
echo "---------------------------"
echo ""

get_char()
{
SAVEDSTTY=`stty -g`
stty -echo
stty cbreak
dd if=/dev/tty bs=1 count=1 2> /dev/null
stty -raw
stty echo
stty $SAVEDSTTY
}
echo "Press any key to start install..."
echo ""
char=`get_char`

echo "---------- Disable SeLinux ----------"

if [ -s /etc/selinux/config ]; then
	sed -i 's/SELINUX=enforcing/SELINUX=disabled/g' /etc/selinux/config
fi

echo "---------- Set Library  ----------"

if [ ! `grep -iqw /lib /etc/ld.so.conf` ]; then
	echo "/lib" >> /etc/ld.so.conf
fi

if [ ! `grep -iqw /usr/lib /etc/ld.so.conf` ]; then
	echo "/usr/lib" >> /etc/ld.so.conf
fi

if [ -d "/usr/lib64" ] && [ ! `grep -iqw /usr/lib64 /etc/ld.so.conf` ]; then
	echo "/usr/lib64" >> /etc/ld.so.conf
fi

if [ ! `grep -iqw /usr/local/lib /etc/ld.so.conf` ]; then
	echo "/usr/local/lib" >> /etc/ld.so.conf
fi

echo "---------- Dependent Packages ----------"

yum -y install gcc gcc-c++ m4 make automake libtool gettext openssl-devel pkgconfig curl curl-devel intltool

echo "================ Transmission Install ==============="

export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig

echo "---------- libevent ----------"

if [ ! -s libevent-*.tar.gz ]; then
	wget -c https://github.com/downloads/libevent/libevent/libevent-2.0.20-stable.tar.gz
fi
tar -zxf libevent-*.tar.gz
cd libevent-*
./configure
make && make install && ldconfig

echo "---------- libiconv ----------"

cd $TRANS_PATH/

if [ ! -s libiconv-*.tar.gz ]; then
	wget -c http://src-mirror.googlecode.com/files/libiconv-1.14.tar.gz
fi
tar -zxf libiconv-*.tar.gz
cd libiconv-*/
./configure --prefix=/usr/local
make && make install && ldconfig

echo "---------- transmission ----------"

cd $TRANS_PATH/

if [ ! -s transmission-*.tar.bz2 ]; then
	wget -c http://download.transmissionbt.com/files/transmission-2.71.tar.bz2
fi
tar -jxf transmission-*.tar.bz2
cd transmission-*

if [ `getconf WORD_BIT` = '32' ] && [ `getconf LONG_BIT` = '64' ] ; then
	./configure --prefix=/usr/local/transmission CFLAGS=-liconv
else
	./configure --prefix=/usr/local/transmission
fi
make && make install

echo "---------- transmission config ----------"

for i in `ls /usr/local/transmission/bin/`; do ln -s /usr/local/transmission/bin/$i /usr/bin/$i; done

cp $TRANS_PATH/init.d.transmission /etc/init.d/transmission
chmod 755 /etc/init.d/transmission
#chkconfig transmission on

mkdir -p $DLDIR > /dev/null 2>&1
mkdir -p /usr/local/transmission/etc/
/etc/init.d/transmission start > /dev/null 2>&1

sed -i 's/240/512/g' /usr/local/transmission/etc/settings.json
sed -i 's/60/128/g' /usr/local/transmission/etc/settings.json
sed -i 's#/root/Downloads#'$DLDIR'#g' /usr/local/transmission/etc/settings.json
sed -i 's/^.*rpc-authentication-required.*/"rpc-authentication-required": true,/' /usr/local/transmission/etc/settings.json
sed -i 's/^.*rpc-password.*/"rpc-password": "'$PASSWORD'",/' /usr/local/transmission/etc/settings.json
sed -i 's/^.*rpc-username.*/"rpc-username": "'$USERNAME'",/' /usr/local/transmission/etc/settings.json
sed -i 's/^.*rpc-whitelist-enabled.*/"rpc-whitelist-enabled": false,/' /usr/local/transmission/etc/settings.json

\cp /usr/local/transmission/etc/settings.json /usr/local/transmission/etc/settings.json.bak

/etc/init.d/transmission restart

service iptables stop > /dev/null 2>&1

clear
echo ""
echo "===================== Install completed ====================="
echo ""
echo "Transmission install completed!"
echo "Intro http://wangyan.org/blog/transmission.html"
echo ""
echo "Transmission Config: /usr/local/transmission/etc/settings.json.bak"
echo "Transmission Web Url: http://IP_ADDRESS:9091"
echo ""
echo "============================================================="
echo ""
